# 01. Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# 02. Import Data

In [2]:
path = r'C:\Users\byl19\Downloads\02-2023 Instacart Basket Analysis'

In [3]:
ords_prods_merge = pd.read_pickle(r'C:\Users\byl19\Downloads\02-2023 Instacart Basket Analysis\02 Data\Prepared Data\opm.pkl')

In [4]:
ords_prods_merge['user_id'].dtype

dtype('O')

In [5]:
ords_prods_merge.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    -1.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  prices    price_range_loc    busiest days  \
0         Soda  ...     9.0  Mid range product  Regularly busy   
1         Soda  ...     9.0  Mid range product    Slowest days   
2         Soda  ...     9.0  Mid range product    Slowest days   
3         Soda  ...     9.0  Mid range product    Slowest days   
4         Soda  ...     9.0  Mid range product    Slowest days   

  busiest_period_of_day max_order   loyalty_flag  average_price spending_flag  \
0        Average orders        10  New custsomer       6.367797   Low spender   
1        Average orders        10  New custsomer       6.367797   Low spender   
2           Most orders        10  New custsomer       6.367797   Low spender   
3        Average orders        10  New custsomer       6.367797   Low spender   
4           Most orders        10  New custsomer       6.367797   Low spender   

   median_days   order_freq_flag  
0         20.0  Regular customer  
1         20.0  Regular customer  
2         20.0  Regular customer  
3         20.0  Regular customer  
4         20.0  Regular customer  

[5 rows x 22 columns]

In [6]:
ords_prods_merge.shape

(32404859, 22)

# 03. Exercise Questions

In [7]:
#2 
ords_prods_merge.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

#3
There is quite a difference when comparing the entire data set to just the subset. The main difference is in the subset there were many missing department id's while the entire data frame includes all 21 of them. When comparing the department id numbers in the subset to the entire there are also some slight differences like department 13 in the subset had a mean of 17.9 and in the entire its 16.5.

In [8]:
#4 loyalty flag transform
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [9]:
#4 loyalty flag loc
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.lo`c[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New custsomer'

In [11]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New custsomer        6243990
Name: loyalty_flag, dtype: int64

In [10]:
#5
ords_prods_merge.groupby('loyalty_flag').agg({'prices':['mean','max','min']})

prices              
                       mean      max  min
loyalty_flag                             
Loyal customer    10.386336  99999.0  1.0
New custsomer     13.294670  99999.0  1.0
Regular customer  12.495717  99999.0  1.0

In [11]:
#6 spending flag transform
ords_prods_merge['average_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [12]:
#6 checking head
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     -1.0         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   

   product_name  ...  department_id  prices  _merge    price_range_loc  \
0          Soda  ...              7     9.0    both  Mid range product   
1          Soda  ...              7     9.0    both  Mid range product   
2          Soda  ...              7     9.0    both  Mid range product   
3          Soda  ...              7     9.0    both  Mid range product   
4          Soda  ...              7     9.0    both  Mid range product   
5          Soda  ...              7     9.0    both  Mid range product   
6          Soda  ...              7     9.0    both  Mid range product   
7          Soda  ...              7     9.0    both  Mid range product   
8          Soda  ...              7     9.0    both  Mid range product   
9          Soda  ...              7     9.0    both  Mid range product   
10         Soda  ...              7     9.0    both  Mid range product   
11         Soda  ...              7     9.0    both  Mid range product   
12         Soda  ...              7     9.0    both  Mid range product   
13         Soda  ...              7     9.0    both  Mid range product   
14         Soda  ...              7     9.0    both  Mid range product   

       busiest day    busiest days busiest_period_of_day max_order  \
0   Regularly busy  Regularly busy        Average orders        10   
1   Regularly busy    Slowest days        Average orders        10   
2   Regularly busy    Slowest days           Most orders        10   
3       Least busy    Slowest days        Average orders        10   
4       Least busy    Slowest days           Most orders        10 

In [20]:
ords_prods_merge.drop(columns=['averge_price', 'averege_price'])

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2398795        1             2                   3   
2           473747        1             3                   3   
3          2254736        1             4                   4   
4           431534        1             5                   4   
...            ...      ...           ...                 ...   
32404854   1320836   202557            17                   2   
32404855     31526   202557            18                   5   
32404856    758936   203436             1                   2   
32404857   2745165   203436             2                   3   
32404858   3093936   205420             1                   4   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                    -1.0         196   
1                         7                    15.0         196   
2                        12                    21.0         196   
3                         7                    29.0         196   
4                        15                    28.0         196   
...                     ...                     ...         ...   
32404854                 15                     1.0       43553   
32404855                 11                     3.0       43553   
32404856                  7                    -1.0       42338   
32404857                  5                    15.0       42338   
32404858                 14                    -1.0       28818   

          add_to_cart_order  reordered                   product_name  ...  \
0                         1          0                           Soda  ...   
1                         1          1                           Soda  ...   
2                         1          1                           Soda  ...   
3                         1          1                           Soda  ...   
4                         1          1                           Soda  ...   
...                     ...        ...                            ...  ...   
32404854                  2          1            Orange Energy Shots  ...   
32404855                  2          1            Orange Energy Shots  ...   
32404856                  4          0          Zucchini Chips, Pesto  ...   
32404857                 16          1          Zucchini Chips, Pesto  ...   
32404858                  8          0  Hot Oatmeal Multigrain Raisin  ...   

          prices  _merge    price_range_loc     busiest day    busiest days  \
0            9.0    both  Mid range product  Regularly busy  Regularly busy   
1            9.0    both  Mid range product  Regularly busy    Slowest days   
2            9.0    both  Mid range product  Regularly busy    Slowest days   
3            9.0    both  Mid range product      Least busy    Slowest days   
4            9.0    both  Mid range product      Least busy    Slowest days   
...          ...     ...                ...             ...             ...   
32404854     3.7    both  Low range product  Regularly busy  Regularly busy   
32404855     3.7    both  Low range product  Regularly busy  Regularly busy   
32404856     6.9    both  Mid range product  Regularly busy  Regularly busy   
32404857     6.9    both  Mid range product  Regularly busy    Slowest days   
32404858    10.3    both  Mid range product      Least busy    Slowest days   

         busiest_period_of_day max_order      loyalty_flag  average_price  \
0               Average orders        10     New custsomer       6.367797   
1               Average orders        10     New custsomer       6.367797   
2                  Most orders        10     New custsomer       6.367797   
3               Average orders        10     New custsomer       6.367797   
4                  Most orders        10     New custsomer       6.367797   
...                        ...       ...               ...            ...   

In [13]:
#6 spending flag loc
ords_prods_merge.loc[ords_prods_merge['average_price'] < 10, 'spending_flag'] = 'Low spender'
ords_prods_merge.loc[ords_prods_merge['average_price'] >= 10, 'spending_flag'] = 'High spender'

In [14]:
#6 check
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

In [15]:
#7 order frequncy flag transform
ords_prods_merge['median_days'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [16]:
#7 check
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     -1.0         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   

   product_name  ...  _merge    price_range_loc     busiest day  \
0          Soda  ...    both  Mid range product  Regularly busy   
1          Soda  ...    both  Mid range product  Regularly busy   
2          Soda  ...    both  Mid range product  Regularly busy   
3          Soda  ...    both  Mid range product      Least busy   
4          Soda  ...    both  Mid range product      Least busy   
5          Soda  ...    both  Mid range product  Regularly busy   
6          Soda  ...    both  Mid range product  Regularly busy   
7          Soda  ...    both  Mid range product  Regularly busy   
8          Soda  ...    both  Mid range product  Regularly busy   
9          Soda  ...    both  Mid range product      Least busy   
10         Soda  ...    both  Mid range product  Regularly busy   
11         Soda  ...    both  Mid range product  Regularly busy   
12         Soda  ...    both  Mid range product  Regularly busy   
13         Soda  ...    both  Mid range product  Regularly busy   
14         Soda  ...    both  Mid range product  Regularly busy   

      busiest days busiest_period_of_day max_order      loyalty_flag  \
0   Regularly busy        Average orders        10     New custsomer   
1     Slowest days        Average orders        10     New custsomer   
2     Slowest days           Most orders        10     New custsomer   
3     Slowest days        Average orders        10     New custsomer   
4     Slowest days           Most orders        10     New custsomer   
5   Regularly busy        Average orders        10     New custsomer   
6     Busiest days       

In [17]:
#7 order frequency loc
ords_prods_merge.loc[ords_prods_merge['median_days'] > 20, 'order_freq_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['median_days'] > 10) & (ords_prods_merge['median_days'] <= 20), 'order_freq_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['median_days'] <= 10, 'order_freq_flag'] = 'Frequent customer'

In [18]:
#7 check
ords_prods_merge[['user_id','order_freq_flag','median_days']].head(60)

user_id        order_freq_flag  median_days
0         1       Regular customer         20.0
1         1       Regular customer         20.0
2         1       Regular customer         20.0
3         1       Regular customer         20.0
4         1       Regular customer         20.0
5         1       Regular customer         20.0
6         1       Regular customer         20.0
7         1       Regular customer         20.0
8         1       Regular customer         20.0
9         1       Regular customer         20.0
10       15      Frequent customer          8.0
11       15      Frequent customer          8.0
12       15      Frequent customer          8.0
13       15      Frequent customer          8.0
14       15      Frequent customer          8.0
15       19      Frequent customer          6.0
16       19      Frequent customer          6.0
17       19      Frequent customer          6.0
18       21      Frequent customer          7.0
19       31      Frequent customer          6.0
20       31      Frequent customer          6.0
21       43      Frequent customer         10.0
22       43      Frequent customer         10.0
23       52      Frequent customer          8.0
24       52      Frequent customer          8.0
25       52      Frequent customer          8.0
26       52      Frequent customer          8.0
27       52      Frequent customer          8.0
28       52      Frequent customer          8.0
29       52      Frequent customer          8.0
30       52      Frequent customer          8.0
31       52      Frequent customer          8.0
32       52      Frequent customer          8.0
33       52      Frequent customer          8.0
34       52      Frequent customer          8.0
35       52      Frequent customer          8.0
36       52      Frequent customer          8.0
37       67      Frequent customer          7.0
38       67      Frequent customer          7.0
39       67      Frequent customer          7.0
40       67      Frequent customer          7.0
41       67      Frequent customer          7.0
42       67      Frequent customer          7.0
43       67      Frequent customer          7.0
44       67      Frequent customer          7.0
45       67      Frequent customer          7.0
46       67      Frequent customer          7.0
47       67      Frequent customer          7.0
48       67      Frequent customer          7.0
49       67      Frequent customer          7.0
50       67      Frequent customer          7.0
51       67      Frequent customer          7.0
52       67      Frequent customer          7.0
53       67      Frequent customer          7.0
54       67      Frequent customer          7.0
55       67      Frequent customer          7.0
56       81  Non-frequent customer         30.0
57       81  Non-frequent customer         30.0
58       82       Regular customer         11.0
59       82       Regular customer         11.0

In [19]:
ords_prods_merge['order_freq_flag'].value_counts(dropna = False)

Frequent customer        22797115
Regular customer          6921016
Non-frequent customer     2686728
Name: order_freq_flag, dtype: int64

In [20]:
#8
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'opm.pkl'))